File: advent_of_code/04.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

func _04() {
	lines := getLines("04.txt")

	getPair := func(raw string) []int {
		var pair []int
		for _, s := range strings.Split(raw, "-") {
			n, err := strconv.Atoi(s)
			if err != nil {
				panic(err)
			}
			pair = append(pair, n)
		}
		return pair
	}

	part1 := func() {
		num := 0
		for _, line := range lines {
			pair := strings.Split(line, ",")
			pair1 := getPair(pair[0])
			pair2 := getPair(pair[1])
			if pair1[0] > pair2[0] {
				if pair1[1] > pair2[1] {
					continue
				}
			} else if pair1[0] < pair2[0] {
				if pair1[1] < pair2[1] {
					continue
				}
			}
			num += 1
		}

		fmt.Println(num)
	}

	part2 := func() {
		num := 0
		for _, line := range lines {
			pair := strings.Split(line, ",")
			pair1 := getPair(pair[0])
			pair2 := getPair(pair[1])
			if pair1[0] > pair2[0] {
				if pair2[1] < pair1[0] {
					continue
				}
			} else if pair1[0] < pair2[0] {
				if pair1[1] < pair2[0] {
					continue
				}
			}
			num += 1
		}

		fmt.Println(num)
	}

	part1()
	part2()
}

File: advent_of_code/2023/04.py

with open('04.txt') as fh:
    txt = fh.read()

def part1():
    def score(line):
        x = line.split(": ")[1].split(" | ")
        y = x[0].split(" ")
        count = len([i for i in x[1].split(" ") if i in y and i])
        if count == 0:
            return 0
        return 2**(count-1)

    print(sum([score(line) for line in txt.split("\n")]))

def part2():
    from collections import defaultdict
    deck = defaultdict(int)
    for i, line in enumerate(txt.split("\n")):
        deck[i] += 1
        x = line.split(": ")[1].split(" | ")
        y = x[0].split(" ")
        count = len([i for i in x[1].split(" ") if i in y and i])
        for j in range(count):
            deck[i+j+1] += deck[i]
    print(sum(deck.values()))

part2()
